src/main/kotlin/cacophony/parser/LLOneParser.kt

```diff
@@ -1,15 +1,79 @@
+        // nextAction gives symbol on the production that we should use next
+        // depending on the current input symbol and dfa state.
+        // null denotes no suitable production, then we should try to accept.
+        // Throws if grammar is not LL(1).
+        inline fun <StateType, reified SymbolType : Enum<SymbolType>> fromAnalyzedGrammar(
             analyzedGrammar: AnalyzedGrammar<StateType, SymbolType>,
-        ): LLOneParser<SymbolType> {
-            TODO("Not yet implemented")
+        ): LLOneParser<StateType, SymbolType> {
+            // Currently the logic is as follows:
+            // 1. Loop over all DFAStateReference and inputSymbol: SymbolType.
+            // 2. For each production with some symbol P, using nullable, first and follow on the start state of the DFA for P
+            //    find all suitable productions (if there is no DFA for P, meaning P is terminal, production is suitable iff P == inputSymbol).
+            // 3. If there are at least 2 suitable productions then our grammar is not LL(1).
+            // 4. If there is exactly one then ok.
+            // 5. If there are none, in process() we should try to accept in current DFA, else it is a parsing error.
+            val nextAction: Map<SymbolType, MutableMap<DFAStateReference<StateType, SymbolType, Production<SymbolType>>, SymbolType?>> =
+                enumValues<SymbolType>().toList().associateWith { mutableMapOf() }
+
+            analyzedGrammar.automata.forEach { (dfaLabel, dfa) -> // automaton for the grammar symbol
+                dfa.getAllStates().forEach { curState -> // dfa state we are currently in
+                    val curStateRef = DFAStateReference(curState, dfa)
+
+                    enumValues<SymbolType>().forEach { inputSymbol -> // symbol from the input
+                        val suitableSymbols = mutableListOf<SymbolType>()
+                        enumValues<SymbolType>().forEach prod@{ prodSymbol -> // symbol in the production
+                            dfa.getProduction(curState, prodSymbol) ?: return@prod
+
+                            val dfaForProdSymbol = analyzedGrammar.automata[prodSymbol]
+                            if (dfaForProdSymbol != null) { // nonterminal
+                                val nextState = dfaForProdSymbol.getStartingState()
+                                val nextStateRef = DFAStateReference(nextState, dfaForProdSymbol)
+                                if ((analyzedGrammar.first[nextStateRef]?.contains(inputSymbol) == true) or (prodSymbol == inputSymbol)) {
+                                    suitableSymbols.add(prodSymbol)
+                                } else if (analyzedGrammar.nullable.contains(nextStateRef) and
+                                    (analyzedGrammar.follow[nextStateRef]?.contains(inputSymbol) == true)
+                                ) {
+                                    suitableSymbols.add(prodSymbol)
+                                }
```

Comment:
no need for `== true`

Reply:
Changed `?` to `!!` and now it is not necessary. Earlier it was, because the type was `Boolean?`.